llvm/include/llvm/Transforms/Utils/SYCLModuleSplit.h

```diff
@@ -0,0 +1,224 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Functionality to split a module into call graphs. A callgraph here is a set
+// of entry points with all functions reachable from them via a call. The result
+// of the split is new modules containing corresponding callgraph.
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SYCL_MODULE_SPLIT_H
+#define LLVM_SYCL_MODULE_SPLIT_H
+
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/Error.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+// TODO(maksimsab):
+//  * check GenXSPIRVWriterAdaptor comments
+//  * Maybe fix doxygen comments.
+
+namespace llvm {
+
+class Function;
+class Module;
+
+enum IRSplitMode {
+  SPLIT_PER_TU,     // one module per translation unit
+  SPLIT_PER_KERNEL, // one module per kernel
+  SPLIT_AUTO,       // automatically select split mode
+  SPLIT_NONE        // no splitting
+};
+
+// \returns IRSplitMode value if \p S is recognized. Otherwise, std::nullopt is
+// returned.
+std::optional<IRSplitMode> convertStringToSplitMode(StringRef S);
+
+// A vector that contains all entry point functions in a split module.
+using EntryPointSet = SetVector<Function *>;
+
+// enum class SyclEsimdSplitStatus { SYCL_ONLY, ESIMD_ONLY, SYCL_AND_ESIMD };
+
+// Describes scope covered by each entry in the module-entry points map
+// populated by the groupEntryPointsByScope function.
+enum EntryPointsGroupScope {
+  Scope_PerKernel, // one entry per kernel
+  Scope_PerModule, // one entry per module
+  Scope_Global     // single entry in the map for all kernels
+};
+
+// Represents a named group of device code entry points - kernels and
+// SYCL_EXTERNAL functions.
+struct EntryPointGroup {
+  // Properties an entry point (EP) group
+  struct Properties {
+    // Scope represented by EPs in a group
+    EntryPointsGroupScope Scope = Scope_Global;
+  };
+
+  std::string GroupId;
+  EntryPointSet Functions;
+  Properties Props;
+
+  EntryPointGroup(StringRef GroupId = "") : GroupId(GroupId) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet Functions)
+      : GroupId(GroupId), Functions(std::move(Functions)) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet Functions,
+                  const Properties &Props)
+      : GroupId(GroupId), Functions(std::move(Functions)), Props(Props) {}
+
+  void saveNames(std::vector<std::string> &Dest) const;
+  void rebuildFromNames(const std::vector<std::string> &Names, const Module &M);
+  void rebuild(const Module &M);
+};
+
+using EntryPointGroupVec = std::vector<EntryPointGroup>;
```

Comment:
What is our justification for using `std::vector` here? [Programmer's Manual](https://llvm.org/docs/ProgrammersManual.html) is pretty clear that when you are in need of a sequential container, then `std::vector` should be used as a last resort when no other LLVM container satisfy your needs